src/gas-price/gas-price.ts

```diff
@@ -0,0 +1,185 @@
+    // eslint-disable-next-line no-console
+    console.log(`Failed to get provider gas price. Error: ${goGasPrice.error.message}.`);
+    throw goGasPrice.error;
+  }
+
+  // Save the new provider recommended gas price to the state
+  setStoreGasPrices(chainId, goGasPrice.data);
+
+  return goGasPrice.data;
+};
+
+/**
+ * Fetches the provider recommended gas price and saves it in the store. Clears out expired gas prices and calculates the gas price to be used in a transaction based on sanitization and scaling settings.
+ * @param chainId
+ * @param rpcUrl
+ * @param gasSettings
+ * @param nonce
+ * @returns {ethers.BigNumber}
+ */
+export const airseekerV2ProviderRecommendedGasPrice = async (
+  chainId: string,
+  rpcUrl: string,
+  gasSettings: GasSettings,
+  nonce?: number
+): Promise<ethers.BigNumber> => {
+  const {
+    recommendedGasPriceMultiplier,
+    sanitizationSamplingWindow,
+    sanitizationPercentile,
+    scalingWindow,
+    scalingMultiplier,
+  } = gasSettings;
+
+  // Initialize the gas store for the chain if not already present
+  if (!gasPriceStore[chainId])
+    gasPriceStore[chainId] = {
+      gasPrices: [],
+      lastUpdateTimestamp: 0,
+      lastUpdateNonce: 0,
+    };
+
+  // Clear expired gas prices from the store
+  clearExpiredStoreGasPrices(chainId, sanitizationSamplingWindow);
+
+  // Get the configured percentile of historical gas prices before adding the new price
+  const percentileGasPrice = getPercentile(
+    sanitizationPercentile,
+    gasPriceStore[chainId]!.gasPrices.map((gasPrice) => gasPrice.price)
+  );
+
+  const gasPrice = await updateGasPriceStore(chainId, rpcUrl);
+
+  // Check if the next update is a retry of a pending transaction and if it has been pending longer than scalingWindow
+  if (
+    nonce &&
+    gasPriceStore[chainId]!.lastUpdateNonce === nonce &&
+    gasPriceStore[chainId] &&
+    gasPriceStore[chainId]!.lastUpdateTimestamp < Date.now() - scalingWindow * 60 * 1_000
+  ) {
+    return multiplyGasPrice(gasPrice, scalingMultiplier);
+  }
+
+  // Check if the multiplied gas price is within the percentile and return the smaller value
+  // TODO should we check for a minimum length of state gas prices used in the calculation?
+  const sanitizedGasPrice = percentileGasPrice && gasPrice.gt(percentileGasPrice) ? percentileGasPrice : gasPrice;
+
+  // const multipliedGasPrice = multiplyGasPrice(sanitizedGasPrice, recommendedGasPriceMultiplier);
+  return multiplyGasPrice(sanitizedGasPrice, recommendedGasPriceMultiplier);
+};
+
+export const runGasPriceCollector = async () => {
+  const config = await loadConfig();
+
+  // TODO: new config gasFetchInterval? Global or chain level?
+  const fetchInterval = /*config.fetchInterval*/ 30 * 1_000;
+
+  if (!gasPriceCollectorInterval) {
+    gasPriceCollectorInterval = setInterval(runGasPriceCollector, fetchInterval);
+  }
+
+  await Promise.all(
+    Object.entries(config.chains).map(([chainId, chain]) =>
+      go(
+        async () =>
```

Comment:
AFAIK this is no longer necessary because the new go utils returns the error correctly, might be worth double-checking that.